### ISSUE TITLE
Cannot check source on stdin

### ISSUE THREAD

**Issue by drmikehenry** [zheller/flake8-quotes#12]
```flake8-quotes``` fails with an exception when attempting to check source files provided on stdin.
```shell
$ git clone git@github.com:zheller/flake8-quotes.git
$ cd flake8-quotes
$ mkvirtualenv flake8-quotes
$ pip install flake8
# This will work (before flake8-quotes is installed):
$ cat flake8_quotes.py | flake8 -
stdin:56:80: E501 line too long (93 > 79 characters)
stdin:59:80: E501 line too long (100 > 79 characters)
stdin:62:80: E501 line too long (84 > 79 characters)
stdin:65:80: E501 line too long (100 > 79 characters)
stdin:81:80: E501 line too long (86 > 79 characters)
# But it will fail with flake8-quotes installed:
$ pip install -e .
(flake8-quotes)[mike@casey:flake8-quotes]$ cat flake8_quotes.py | flake8 -
Traceback (most recent call last):
  File "/home/mike/envs/flake8-quotes/bin/flake8", line 11, in <module>
    sys.exit(main())
  File "/home/mike/envs/flake8-quotes/lib/python3.4/site-packages/flake8/main.py", line 33, in main
    report = flake8_style.check_files()
  File "/home/mike/envs/flake8-quotes/lib/python3.4/site-packages/flake8/engine.py", line 176, in check_files
    return self._retry_serial(self._styleguide.check_files, paths=paths)
  File "/home/mike/envs/flake8-quotes/lib/python3.4/site-packages/flake8/engine.py", line 167, in _retry_serial
    return func(*args, **kwargs)
  File "/home/mike/envs/flake8-quotes/lib/python3.4/site-packages/pep8.py", line 1672, in check_files
    runner(path)
  File "/home/mike/envs/flake8-quotes/lib/python3.4/site-packages/flake8/engine.py", line 121, in input_file
    return fchecker.check_all(expected=expected, line_offset=line_offset)
  File "/home/mike/envs/flake8-quotes/lib/python3.4/site-packages/pep8.py", line 1412, in check_all
    self.check_ast()
  File "/home/mike/envs/flake8-quotes/lib/python3.4/site-packages/pep8.py", line 1359, in check_ast
    for lineno, offset, text, check in checker.run():
  File "/home/mike/build/flake8-quotes/flake8_quotes.py", line 51, in run
    noqa_line_numbers = self.get_noqa_lines(file_contents)
  File "/home/mike/build/flake8-quotes/flake8_quotes.py", line 59, in get_noqa_lines
    tokens = [Token(t) for t in tokenize.generate_tokens(lambda L=iter(file_contents): next(L))]
ValueError: I/O operation on closed file.
```

**Comment by twolfson** [comment]
Yikes, thanks for the bug report. We will take a look at this by the end of the weekend.

**Comment by twolfson** [comment]
Looking into triaging this now

**Comment by twolfson** [comment]
Alright, it looks like it was silently swallowing errors on Python 2.7 and I have reproduced the Python 3.4 error as well.

**Comment by twolfson** [comment]
It looks like the issue is being caused by someone else consuming `stdin` and using `stdin.close()` before we can read it.

```python
$ echo 'hi' | python -c "from sys import stdin; print(stdin.read()); print(stdin.read())"
hi


(flake8-quotes3)todd at Euclid in ~/github/flake8-quotes on bug/stdin.flake8
$ echo 'hi' | python -c "from sys import stdin; print(stdin.read()); stdin.close(); print(stdin.read())" 
hi

Traceback (most recent call last):
  File "<string>", line 1, in <module>
ValueError: I/O operation on closed file.
```

We could explore converting the AST back into standard form but I think that's counter productive.

As a result, I started exploring the other packages to see what they but have came up empty (there was a `pep8.stdin_get_value()` but that is broken)

https://github.com/public/flake8-import-order/blob/2ac7052a4e02b4a8a0125a106d87465a3b9fd688/flake8_import_order/__init__.py#L211

However, I did find an alternate form to accepting content from `pep8`/`flake8`. It seems to be that if we pass in a `function` that we can get the lines of the file:

https://github.com/PyCQA/pep8/blob/4dc42d842274ba27d2724e76eb83ff69e7db226f/pep8.py#L1334-L1342

Since this is undocumented, I will be opening an issue to make sure it's safe to use.

**Comment by twolfson** [comment]
I have opened an issue in the respective GitLab repo:

https://gitlab.com/pycqa/flake8/issues/102

**Comment by drmikehenry** [comment]
Thanks for looking into this so quickly! :-)

**Comment by zedlander** [comment]
@twolfson Why do you say `pep8.stdin_get_value()` is broken? I have it working in 2.7 and 3.5 locally. Can push a PR if you want.

**Comment by twolfson** [comment]
It's been a while since I last touched this issue but I believe other plugins were consuming the `stdin` content and the buffer would be seeked to EOF. However, if you have something working, I would love to see a PR =)

**Comment by zedlander** [comment]
Done :) https://github.com/zheller/flake8-quotes/pull/16

**Comment by twolfson** [comment]
Wow, very speedy. Nicely done :+1:

**Comment by twolfson** [comment]
This has been resolved by @zedlander in #16 and released in `0.2.0`.
